waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,585 @@
+-- reports an error, without consuming the next non-trivia token.
+parsePrimitive :: TokenKind -> ParserM ()
+parsePrimitive kind =
+  (consumeWhitespace >> peek) >>= \case
+    Just nextToken | tokenKind nextToken == kind -> do
+      advance
+      let newNode =
+            SyntaxNode
+              { snodeKind = S.Token kind,
+                snodeWidth = tokenWidth nextToken,
+                snodeChildren = []
+              }
+      pushNode newNode
+    nextToken -> makeError nextToken (TokenSet.fromKind kind)
+
+-- | Parse "EOF" rule. If there are no non-trivia tokens in the remaining input,
+-- this parser succeeds without making any nodes. Otherwise, it consumes all
+-- remaining input and reports an error at the first non-trivia token.
+parseEOF :: ParserM ()
+parseEOF =
+  (consumeWhitespace >> peek) >>= \case
+    Just nextToken -> do
+      startOffset <- gets pstateNextOffset
+      advance
+      let firstTokenKind = tokenKind nextToken
+      -- TODO: consider reporting all kinds here? not sure
+      _tokenKinds <- collectUntilEOF
+      throwError $
+        ParseError $
+          Unexpected
+            (Span startOffset (startOffset + tokenWidth nextToken))
+            firstTokenKind
+            TokenSet.fromEOF
+      where
+        collectUntilEOF :: ParserM [TokenKind]
+        collectUntilEOF =
+          peek >>= \case
+            Just next -> advance >> ((tokenKind next :) <$> collectUntilEOF)
+            Nothing -> pure []
+    Nothing -> do
+      return ()
+
+-- | Parse "As" rule. Identical to "Primitive", but uses the label as the kind
+-- for the syntax node when it succeeds.
+parseAs :: SyntaxKind -> TokenKind -> ParserM ()
+parseAs label kind =
+  (consumeWhitespace >> peek) >>= \case
+    Just nextToken | tokenKind nextToken == kind -> do
+      advance
+      let newNode =
+            SyntaxNode
+              { snodeKind = label,
+                snodeWidth = tokenWidth nextToken,
+                snodeChildren = []
+              }
+      pushNode newNode
+    nextToken -> makeError nextToken (TokenSet.fromKind kind)
+
+-- | Parse "Group" rule. Runs the inner parser. When it succeeds, it puts the
+-- outputted nodes from that parser under a parent node with the label. When it
+-- fails, it recovers from the error by catching the error or unwinding.
+parseGroup :: SyntaxKind -> GrammarRule -> ParserM ()
+parseGroup label inner = do
+  let pushGroupNode children =
+        let width = sum $ map snodeWidth children
+         in pushNode $
+              SyntaxNode
+                { snodeKind = label,
+                  snodeWidth = width,
+                  snodeChildren = children
+                }
+
+  consumeWhitespace
+  siblingNodes <- gets pstateNodes
+  modify (\s -> s {pstateNodes = []})
+  result <- tryError $ doParse inner
+  childNodes <- gets pstateNodes
+  modify (\s -> s {pstateNodes = siblingNodes})
+  case result of
+    Nothing -> do
+      pushGroupNode childNodes
+    Just err -> do
```

Comment:
Ok, so here starts error handling, we certainly should make this a separate function, either standalone or in `where`, I am not sure yet.